frontend/src/app/chart/chart.component.ts

```diff
@@ -27,55 +27,72 @@ import { HttpResponse } from '@angular/common/http';
 })
 export class ChartComponent implements OnInit {
   // Data related variable
-  records: Record[];
   loading = false;
+  number = 600;
+  records: Record[];
+  strategy = 'max';
+  zoomIn = false;
 
   // Chart d3 SVG Elements
+  private brush: d3.BrushBehavior<unknown>;
+  private line: d3.Line<[number, number]>;
+  private svg: d3.Selection<SVGGElement, unknown, HTMLElement, any>;
+  private svgChart: d3.Selection<SVGGElement, unknown, HTMLElement, any>;
   private xAxis: d3.Selection<SVGGElement, unknown, HTMLElement, any>;
   private xScale: d3.ScaleTime<number, number>;
+  private yAxis: d3.Selection<SVGGElement, unknown, HTMLElement, any>;
   private yScale: d3.ScaleLinear<number, number>;
-  private svgChart: d3.Selection<SVGGElement, unknown, HTMLElement, any>;
-  private brush: d3.BrushBehavior<unknown>;
 
   // Chart size constants
-  chartWidth = 1100;
-  chartHeight = 700;
+  chartHeight = 400;
   chartPadding = 50;
-  timeParse = d3.timeParse('%s');
-  timeFormat = d3.timeFormat('%m/%d-%H:%M');
+  chartWidth = 1100;
+  timeParse = d3.timeParse('%Q');
+  timeFormat = d3.timeFormat('%H:%M:%S.%L');
+  totalXDomain: Date[];
+  animationDuration = 500;
 
-  constructor(private service: GetDataService) {}
+  constructor(private service: HttpService) {}
 
   ngOnInit(): void {
+    this.loadRecords(this.initChart.bind(this));
+  }
+
+  preprocess() {
+    console.log('preprocessing');
+  }
+
+  loadRecords(chartFunction: () => void, timespan?: Date[]) {
```

Comment:
Passing initChart and updateChartDomain in loadRecord is to load data before update the chart. Because data loading requires async operation, simply calling initChart and updateChartDomain would likely work on outdated or empty data instead.

Reply:
You can call the method inside your subscribe method so that it will be triggered after you load the data from the backend. It's a bit readability unfriendly for your current code structure.

Reply:
Also remember to unsubscribe your observables at ngOnDestory to prevent possible memory leak.

Reply:
fixed both unsubscribe and chartFunction. Not sure if it's the right way to unsubscribe, please check it.